servlet/src/tomcatTest/java/io/grpc/servlet/TomcatInteropTest.java

```diff
@@ -0,0 +1,166 @@
+  public void before() {
+    Logger rootLogger = LogManager.getLogManager().getLogger("");
+    // rootLogger.setLevel(Level.ALL);
+    for (Handler h : rootLogger.getHandlers()) {
+      h.setLevel(Level.FINEST);
+    }
+    Logger.getLogger(ServletServerStream.class.getName()).setLevel(Level.FINEST);
+    Logger.getLogger(AsyncServletOutputStreamWriter.class.getName()).setLevel(Level.FINEST);
+  }
+
+  @After
+  @Override
+  public void tearDown() {
+    super.tearDown();
+    try {
+      server.stop();
+    } catch (LifecycleException e) {
+      throw new AssertionError(e);
+    }
+  }
+
+  @AfterClass
+  public static void cleanUp() throws Exception {
+    FileUtils.deleteDirectory(new File("tomcat.0"));
+  }
+
+  @Override
+  protected ServerBuilder<?> getServerBuilder() {
+    return new ServletServerBuilder().maxInboundMessageSize(AbstractInteropTest.MAX_MESSAGE_SIZE);
+  }
+
+  @Override
+  protected void startServer(ServerBuilder<?> builer) {
+    server = new Tomcat();
+    server.setPort(0);
+    Context ctx = server.addContext(MYAPP, new File("build/tmp").getAbsolutePath());
+    Tomcat
+        .addServlet(
+            ctx, "io.grpc.servlet.TomcatInteropTest",
+            new GrpcServlet(((ServletServerBuilder) builer).buildServletAdapter()))
+        .setAsyncSupported(true);
+    ctx.addServletMappingDecoded("/*", "io.grpc.servlet.TomcatInteropTest");
+    Http2Protocol http2Protocol = new Http2Protocol();
+    http2Protocol.setOverheadCountFactor(0);
+    server.getConnector().addUpgradeProtocol(http2Protocol);
+    try {
+      server.start();
+    } catch (LifecycleException e) {
+      throw new RuntimeException(e);
+    }
+
+    port = server.getConnector().getLocalPort();
+  }
+
+  @Override
+  protected ManagedChannelBuilder<?> createChannelBuilder() {
+    NettyChannelBuilder builder =
+            (NettyChannelBuilder) ManagedChannelBuilder.forAddress(HOST, port)
+                    .usePlaintext()
+                    .maxInboundMessageSize(AbstractInteropTest.MAX_MESSAGE_SIZE);
+    InternalNettyChannelBuilder.setStatsEnabled(builder, false);
+    builder.intercept(createCensusStatsClientInterceptor());
+    return builder;
+  }
+
+  @Override
+  protected boolean metricsExpected() {
+    return false; // otherwise re-test will not work
+  }
+
+  // FIXME
+  @Override
+  @Ignore("Tomcat is broken on client GOAWAY")
+  @Test
+  public void gracefulShutdown() {}
+
+  // FIXME
+  @Override
+  @Ignore("Tomcat is not able to send trailer only")
+  @Test
+  public void specialStatusMessage() {}
+
+  // FIXME
+  @Override
+  @Ignore("Tomcat is not able to send trailer only")
+  @Test
+  public void unimplementedMethod() {}
+
+  // FIXME
+  @Override
+  @Ignore("Tomcat is not able to send trailer only")
+  @Test
+  public void statusCodeAndMessage() {}
+
+  // FIXME
+  @Override
+  @Ignore("Tomcat is not able to send trailer only")
+  @Test
+  public void emptyStream() {}
+
+  // Fails intermittently
+  //@Ignore
```

Comment:
At one point it failed intermittently, and then it apparently stopped failing? I haven't had time to figure out why it fails, left the test running so that the next time there might be a chance to figure it out.